doc/subsystems/shell/shell.rst

```diff
@@ -0,0 +1,296 @@
+* :command:`resize` - Must be executed when terminal width is different than 80
+  characters or after each change of terminal width. It ensures proper
+  multiline text display and :kbd:`←`, :kbd:`→`, :kbd:`End`, :kbd:`Home` keys
+  handling. Currently :command:`resize` command works only with UART flow
+  control switched on. It can be also called with a subcommand:
+
+	* :command:`default` - Shell will send send terminal width = 80 to the
+	  terminal and assume successful delivery.
+
+* :command:`shell` - Root command with useful shell-related subcommands like:
+
+	* :command:`echo` - Toggles shell echo.
+        * :command:`colors` - Toggles colored syntax. This might be helpful in
+          case of Bluetooth shell to limit the amount of transferred bytes.
+	* :command:`stats` - Shows shell statistics.
+
+Commands help
+=============
+
+Every user-defined command, subcommand, or option can have its own help
+description. The help for commands and subcommands can be created with
+respective macros: :c:macro:`SHELL_CMD_REGISTER` and :c:macro:`SHELL_CMD`.
+In addition, you can define options for commands or subcommands using the
+macro :c:macro:`SHELL_OPT`. By default, each and every command or subcommand
+has these two options implemented: ``-h`` and ``--help``.
+
+In order to add help functionality to a command or subcommand, you must
+implement the help handler by either calling :cpp:func:`shell_cmd_precheck`
+or pair of functions :cpp:func:`shell_help_requested` and
+:cpp:func:`shell_help_print`. The former is more convenient as it also
+checks for valid arguments count.
+
+
+The following code is an example of how commands can be processed:
+
+.. code-block:: c
+
+	static void cmd_resize(const struct shell *shell, size_t argc,
+				char **argv)
+	{
+		int err;
+
+		if (!IS_ENABLED(CONFIG_SHELL_CMDS_RESIZE)) {
+			shell_fprintf(shell, SHELL_ERROR,
+					"Command not supported.\r\n");
+			return;
+		}
+
+		/* function responsible for printing command help */
+		if (!shell_cmd_precheck(shell, (argc <= 2), NULL, 0)) {
+			return;
+		}
+
+		if (argc != 1) {
+			shell_fprintf(shell, SHELL_ERROR, "%s:%s%s\r\n",
+				      argv[0],
+				      SHELL_MSG_UNKNOWN_PARAMETER, argv[1]);
+			return;
+		}
+
+		err = terminal_size_get(shell);
+		if (err != 0) {
+			shell->ctx->vt100_ctx.cons.terminal_wid =
+					SHELL_DEFAULT_TERMINAL_WIDTH;
+			shell->ctx->vt100_ctx.cons.terminal_hei =
+					SHELL_DEFAULT_TERMINAL_HEIGHT;
+			shell_fprintf(shell, SHELL_WARNING,
+				      "No response from the terminal, assumed"
+					" 80x24 screen size\r\n");
+		}
+	}
+
+When possible, use subcommands instead of options.  Options apply mainly in the
+case when an argument with ``-`` or ``--`` is requested. The main benefit of
+using subcommands is that they can be prompted or completed with the :kbd:`Tab`
+key. In addition, subcommands can have their own handler, which limits the
+usage of ``if - else if`` statements combination with the ``strcmp`` function.
```

Comment:
... function _in command handler_. Missing information where this usage is limited.